diff --git a/CMakeLists.txt b/CMakeLists.txt
index ec87f18f..c28bc4ac 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -154,7 +154,7 @@ include_directories(${PROJECT_SOURCE_DIR}/include)
 include_directories(${PROJECT_BINARY_DIR}/include) # Tablegen'd files
 include_directories(${PROJECT_SOURCE_DIR}/third_party)
 include_directories(${PROJECT_BINARY_DIR}/third_party) # Tablegen'd files
-
+include_directories(${PROJECT_BINARY_DIR}/third_party/spine_triton/include)
 # link_directories(${LLVM_LIBRARY_DIR})
 add_subdirectory(include)
 add_subdirectory(lib)
diff --git a/bin/RegisterTritonDialects.h b/bin/RegisterTritonDialects.h
index 39bf50b7..b438161c 100644
--- a/bin/RegisterTritonDialects.h
+++ b/bin/RegisterTritonDialects.h
@@ -26,6 +26,7 @@
 #include "mlir/Dialect/LLVMIR/NVVMDialect.h"
 #include "mlir/Dialect/LLVMIR/ROCDLDialect.h"
 #include "mlir/InitAllPasses.h"
+#include "../../include/triton-shared/Dialect/smt/IR/SMTDialect.h"
 
 namespace mlir {
 namespace test {
@@ -80,7 +81,8 @@ inline void registerTritonDialects(mlir::DialectRegistry &registry) {
       mlir::triton::nvidia_gpu::TritonNvidiaGPUDialect,
       mlir::triton::gpu::TritonGPUDialect, mlir::math::MathDialect,
       mlir::arith::ArithDialect, mlir::scf::SCFDialect, mlir::gpu::GPUDialect,
-      mlir::LLVM::LLVMDialect, mlir::NVVM::NVVMDialect
+      mlir::LLVM::LLVMDialect, mlir::NVVM::NVVMDialect,
+      mlir::smt::SMTDialect
       // mlir::triton::nvgpu::NVGPUDialect,
       // mlir::triton::nvws::NVWSDialect,
       // mlir::triton::amdgpu::TritonAMDGPUDialect,
diff --git a/python/src/ir.cc b/python/src/ir.cc
index 387f73d1..50c01f8a 100644
--- a/python/src/ir.cc
+++ b/python/src/ir.cc
@@ -38,8 +38,7 @@
 #include "llvm/Support/SourceMgr.h"
 
 #include "third_party/proton/dialect/include/Dialect/Proton/IR/Dialect.h"
-
-namespace {
+#include "../../../include/triton-shared/Dialect/smt/IR/SMTDialect.h"
 
 namespace py = pybind11;
 using namespace mlir;
@@ -47,7 +46,12 @@ using namespace triton;
 namespace tt = triton;
 namespace ttg = triton::gpu;
 namespace ttng = triton::nvidia_gpu;
+namespace ir {
 
+// Pointer to the TritonOpBuilder class, used to register IR ops for third-party
+// dialects.
+static py::class_<TritonOpBuilder> *builderClassPtr = nullptr;
+py::class_<TritonOpBuilder> *getBuilderClass() { return builderClassPtr; }
 llvm::raw_fd_ostream &mlir_dumps() {
   std::error_code EC;
   static llvm::raw_fd_ostream S(::triton::tools::getStrEnv("MLIR_DUMP_PATH"),
@@ -190,13 +194,13 @@ py::list getTensorDescMetadata(ModuleOp &mod) {
   }
   return result;
 }
-
-} // anonymous namespace
+}
 
 /*****************************************************************************/
 /* Python bindings for ir                                                    */
 /*****************************************************************************/
 
+using namespace ir;
 void init_triton_ir(py::module &&m) {
   using ret = py::return_value_policy;
   using namespace pybind11::literals;
@@ -304,7 +308,7 @@ void init_triton_ir(py::module &&m) {
                     math::MathDialect, arith::ArithDialect, scf::SCFDialect,
                     ::mlir::gpu::GPUDialect, cf::ControlFlowDialect,
                     ::mlir::triton::proton::ProtonDialect, LLVM::LLVMDialect,
-                    mlir::ub::UBDialect>();
+                    mlir::ub::UBDialect, mlir::smt::SMTDialect>();
     mlir::LLVM::registerInlinerInterface(registry);
     registerBuiltinDialectTranslation(registry);
     registerLLVMDialectTranslation(registry);
@@ -458,6 +462,7 @@ void init_triton_ir(py::module &&m) {
   py::class_<IntegerAttr, Attribute>(m, "integer_attr", py::module_local());
   py::class_<BoolAttr, Attribute>(m, "bool_attr", py::module_local());
   py::class_<UnitAttr, Attribute>(m, "unit_attr", py::module_local());
+  py::class_<StringAttr, Attribute>(m, "string_attr", py::module_local());
 
   // Ops
   py::class_<OpState>(m, "OpState", py::module_local())
@@ -711,9 +716,13 @@ void init_triton_ir(py::module &&m) {
 
   py::class_<OpBuilder::InsertPoint>(m, "InsertPoint", py::module_local());
 
-  py::class_<TritonOpBuilder>(m, "builder", py::module_local(),
-                              py::dynamic_attr())
-      .def(py::init<MLIRContext *>())
+  // The static builderClass object persists throughout the compilation,
+  // allowing third-party backends to register their ops separately.
+  static py::class_<TritonOpBuilder> builderClass(
+      m, "builder", py::module_local(), py::dynamic_attr());
+  builderClassPtr = &builderClass;
+  builderClass.def(py::init<MLIRContext *>())
+      .def("get_op_builder", &TritonOpBuilder::getBuilder, ret::reference)
       // getters
       .def("create_module",
            [](TritonOpBuilder &self) -> ModuleOp {
diff --git a/python/src/ir.h b/python/src/ir.h
index b27c41a7..2e4d3b31 100644
--- a/python/src/ir.h
+++ b/python/src/ir.h
@@ -2,6 +2,9 @@
 #include "mlir/IR/Builders.h"
 #include "triton/Tools/Sys/GetEnv.hpp"
 #include <memory>
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
 
 // A custom op builder that keeps track of the last location
 class TritonOpBuilder {
@@ -88,3 +91,7 @@ private:
   bool lineInfoEnabled =
       !mlir::triton::tools::getBoolEnv("TRITON_DISABLE_LINE_INFO");
 };
+
+namespace ir {
+extern py::class_<TritonOpBuilder> *getBuilderClass();
+} // namespace ir
\ No newline at end of file
