diff --git a/python/src/ir.cc b/python/src/ir.cc
old mode 100644
new mode 100755
index 387f73d1..9e48a278
--- a/python/src/ir.cc
+++ b/python/src/ir.cc
@@ -39,14 +39,18 @@
 
 #include "third_party/proton/dialect/include/Dialect/Proton/IR/Dialect.h"
 
-namespace {
-
 namespace py = pybind11;
 using namespace mlir;
 using namespace triton;
 namespace tt = triton;
 namespace ttg = triton::gpu;
 namespace ttng = triton::nvidia_gpu;
+namespace ir {
+
+// Pointer to the TritonOpBuilder class, used to register IR ops for third-party
+// dialects.
+static py::class_<TritonOpBuilder> *builderClassPtr = nullptr;
+py::class_<TritonOpBuilder> *getBuilderClass() { return builderClassPtr; }
 
 llvm::raw_fd_ostream &mlir_dumps() {
   std::error_code EC;
@@ -190,13 +194,13 @@ py::list getTensorDescMetadata(ModuleOp &mod) {
   }
   return result;
 }
-
-} // anonymous namespace
+}
 
 /*****************************************************************************/
 /* Python bindings for ir                                                    */
 /*****************************************************************************/
 
+using namespace ir;
 void init_triton_ir(py::module &&m) {
   using ret = py::return_value_policy;
   using namespace pybind11::literals;
@@ -458,6 +462,7 @@ void init_triton_ir(py::module &&m) {
   py::class_<IntegerAttr, Attribute>(m, "integer_attr", py::module_local());
   py::class_<BoolAttr, Attribute>(m, "bool_attr", py::module_local());
   py::class_<UnitAttr, Attribute>(m, "unit_attr", py::module_local());
+  py::class_<StringAttr, Attribute>(m, "string_attr", py::module_local());
 
   // Ops
   py::class_<OpState>(m, "OpState", py::module_local())
@@ -711,9 +716,13 @@ void init_triton_ir(py::module &&m) {
 
   py::class_<OpBuilder::InsertPoint>(m, "InsertPoint", py::module_local());
 
-  py::class_<TritonOpBuilder>(m, "builder", py::module_local(),
-                              py::dynamic_attr())
-      .def(py::init<MLIRContext *>())
+  // The static builderClass object persists throughout the compilation,
+  // allowing third-party backends to register their ops separately.
+  static py::class_<TritonOpBuilder> builderClass(
+      m, "builder", py::module_local(), py::dynamic_attr());
+  builderClassPtr = &builderClass;
+  builderClass.def(py::init<MLIRContext *>())
+      .def("get_op_builder", &TritonOpBuilder::getBuilder, ret::reference)
       // getters
       .def("create_module",
            [](TritonOpBuilder &self) -> ModuleOp {
diff --git a/python/src/ir.h b/python/src/ir.h
old mode 100644
new mode 100755
index b27c41a7..2e4d3b31
--- a/python/src/ir.h
+++ b/python/src/ir.h
@@ -2,6 +2,9 @@
 #include "mlir/IR/Builders.h"
 #include "triton/Tools/Sys/GetEnv.hpp"
 #include <memory>
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
 
 // A custom op builder that keeps track of the last location
 class TritonOpBuilder {
@@ -88,3 +91,7 @@ private:
   bool lineInfoEnabled =
       !mlir::triton::tools::getBoolEnv("TRITON_DISABLE_LINE_INFO");
 };
+
+namespace ir {
+extern py::class_<TritonOpBuilder> *getBuilderClass();
+} // namespace ir
\ No newline at end of file
diff --git a/python/triton/compiler/code_generator.py b/python/triton/compiler/code_generator.py
old mode 100644
new mode 100755
index 1dd7b4ed..c2d3a3e0
--- a/python/triton/compiler/code_generator.py
+++ b/python/triton/compiler/code_generator.py
@@ -1063,7 +1063,8 @@ class CodeGenerator(ast.NodeVisitor):
         disallow_acc_multi_buffer = False
         flatten = False
         warp_specialize = False
-        if IteratorClass is language.range:
+        bind_sub_block = None
+        if IteratorClass in [language.range, language.extra.smt.parallel]:
             iterator = IteratorClass(*iter_args, **iter_kwargs)
             # visit iterator arguments
             # note: only `range` iterator is supported now
@@ -1076,6 +1077,8 @@ class CodeGenerator(ast.NodeVisitor):
             disallow_acc_multi_buffer = iterator.disallow_acc_multi_buffer
             flatten = iterator.flatten
             warp_specialize = iterator.warp_specialize
+            if (IteratorClass is language.extra.smt.parallel):
+                bind_sub_block = iterator.bind_sub_block
         elif IteratorClass is range:
             # visit iterator arguments
             # note: only `range` iterator is supported now
@@ -1132,6 +1135,8 @@ class CodeGenerator(ast.NodeVisitor):
                 for_op.set_attr("tt.flatten", self.builder.get_unit_attr())
             if warp_specialize:
                 for_op.set_attr("tt.warp_specialize", self.builder.get_unit_attr())
+            if (bind_sub_block is not None) and bind_sub_block:
+                for_op.set_attr("bind_sub_block", self.builder.get_bool_attr(bind_sub_block))
 
             self.scf_stack.append(node)
             for_op_body = for_op.get_body(0)
diff --git a/setup.py b/setup.py
old mode 100644
new mode 100755
index 345986fa..08274e9a
--- a/setup.py
+++ b/setup.py
@@ -27,6 +27,7 @@ from setuptools.command.sdist import sdist
 from dataclasses import dataclass
 
 import pybind11
+import glob
 
 try:
     from setuptools.command.bdist_wheel import bdist_wheel
@@ -329,7 +330,7 @@ def download_and_copy(name, src_func, dst_path, variable, version, url_func):
     arch = platform.machine()
     # NOTE: This might be wrong for jetson if both grace chips and jetson chips return aarch64
     arch = {"arm64": "sbsa", "aarch64": "sbsa"}.get(arch, arch)
-    supported = {"Linux": "linux", "Darwin": "linux"}
+    supported = {"Linux": "linux", "Darwin": "macos"}
     url = url_func(supported[system], arch, version)
     src_path = src_func(supported[system], arch, version)
     tmp_path = os.path.join(triton_cache_path, "nvidia", name)  # path to cache the download
@@ -518,8 +519,55 @@ class CMakeBuild(build_ext):
         cmake_dir = get_cmake_dir()
         subprocess.check_call(["cmake", self.base_dir] + cmake_args, cwd=cmake_dir, env=env)
         subprocess.check_call(["cmake", "--build", "."] + build_args, cwd=cmake_dir)
+        self.copy_to_package(cmake_dir)
+
         subprocess.check_call(["cmake", "--build", ".", "--target", "mlir-doc"], cwd=cmake_dir)
 
+    def copy_to_package(self, cmake_dir, file_mappings=None):
+        if file_mappings is None:
+            file_mappings = {
+                "**/spine-triton-opt/spine-triton-opt": "bin",
+            }
+
+        base_target_dir = os.path.join(self.base_dir, "python", "triton", "backends", "spine_triton")
+        copied_files = []
+        file_conflicts = set()
+
+        for pattern, sub_dir in file_mappings.items():
+            target_dir = os.path.join(base_target_dir, sub_dir)
+            os.makedirs(target_dir, exist_ok=True)
+            found_files = glob.glob(os.path.join(cmake_dir, pattern), recursive=True)
+
+            if not found_files:
+                print(f"Warning: No files matching the pattern '{pattern}' were found")
+                continue
+
+            for src_file_path in found_files:
+                file_name = os.path.basename(src_file_path)
+                target_path = os.path.join(target_dir, file_name)
+                if file_name in file_conflicts:
+                    print(f"Warning: The file {file_name} may match multiple patterns, resulting in a path conflict")
+                    continue
+
+                file_conflicts.add(file_name)
+
+                try:
+                    shutil.copy(src_file_path, target_path)
+                    if not file_name.endswith('.whl') and not file_name.endswith('.so'):
+                        os.chmod(target_path, 0o755)
+
+                    copied_files.append((file_name, sub_dir))
+                    print(f"Copied {file_name} to: {target_path}")
+                except Exception as e:
+                    print(f"Error copying file {file_name} to {sub_dir}: {e}")
+
+        if not copied_files:
+            print("Warning: No files were successfully copied")
+        else:
+            print(f"Successfully copied {len(copied_files)} files:")
+            for file_name, sub_dir in copied_files:
+                print(f"  - {file_name} -> {sub_dir}/")
+
 
 def download_and_copy_dependencies():
     return
@@ -621,6 +669,8 @@ def get_package_dirs():
             for x in os.listdir(backend.tools_dir):
                 yield (f"triton.tools.extra.{x}", os.path.join(backend.tools_dir, x))
 
+    yield ("triton.backends.spine_triton.bin", "python/triton/backends/spine_triton/bin")
+
     if check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
         yield ("triton.profiler", "third_party/proton/proton")
 
@@ -643,6 +693,8 @@ def get_packages():
             for x in os.listdir(backend.tools_dir):
                 yield f"triton.tools.extra.{x}"
 
+    yield "triton.backends.spine_triton.bin"
+
     if check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
         yield "triton.profiler"
 
@@ -766,7 +818,7 @@ def get_git_version_suffix():
 
 
 # keep it separate for easy substitution
-TRITON_VERSION = "3.3.0" + get_git_version_suffix() + os.environ.get("TRITON_WHEEL_VERSION_SUFFIX", "")
+TRITON_VERSION = "3.3.0"  + get_git_version_suffix() + os.environ.get("TRITON_WHEEL_VERSION_SUFFIX", "") + "spacemit"
 
 # Dynamically define supported Python versions and classifiers
 MIN_PYTHON = (3, 9)
@@ -797,8 +849,11 @@ setup(
     ],
     packages=list(get_packages()),
     package_dir=dict(get_package_dirs()),
-    entry_points=get_entry_points(),
+    package_data={
+        'triton.backends.spine_triton.bin': ['spine-triton-opt'],
+    },
     include_package_data=True,
+    entry_points=get_entry_points(),
     ext_modules=[CMakeExtension("triton", "triton/_C/")],
     cmdclass={
         "bdist_wheel": plugin_bdist_wheel,
