diff --git a/include/triton/Dialect/Triton/IR/TritonOps.td b/include/triton/Dialect/Triton/IR/TritonOps.td
index 43fbc16a2..238f4d812 100644
--- a/include/triton/Dialect/Triton/IR/TritonOps.td
+++ b/include/triton/Dialect/Triton/IR/TritonOps.td
@@ -1374,5 +1374,22 @@ def TT_DescriptorScatterOp : TT_Op<"descriptor_scatter", [TT_DescriptorStoreLike
   let hasVerifier = 1;
 }
 
+def TT_AnnotationOp : TT_Op<"annotation", [Pure, MemoryEffects<[MemWrite]>]> {
+  let summary = "Annotate a tensor with key-value attribute pairs";
+  let description = [{
+    `tt.annotation` operation can be used to annotate a tensor with
+    key-value attribute pairs.
+
+    Example:
+    ```mlir
+    tt.annotation %target {key : val}
+    ```
+  }];
+  let arguments = (ins TT_Tensor:$src);
+  let assemblyFormat = [{
+    $src attr-dict `:` type($src)
+  }];
+}
+
 
 #endif // Triton_OPS
diff --git a/python/src/ir.cc b/python/src/ir.cc
index 387f73d12..2a3c6ca80 100644
--- a/python/src/ir.cc
+++ b/python/src/ir.cc
@@ -458,6 +458,7 @@ void init_triton_ir(py::module &&m) {
   py::class_<IntegerAttr, Attribute>(m, "integer_attr", py::module_local());
   py::class_<BoolAttr, Attribute>(m, "bool_attr", py::module_local());
   py::class_<UnitAttr, Attribute>(m, "unit_attr", py::module_local());
+  py::class_<StringAttr, Attribute>(m, "string_attr", py::module_local());
 
   // Ops
   py::class_<OpState>(m, "OpState", py::module_local())
@@ -1687,6 +1688,13 @@ void init_triton_ir(py::module &&m) {
       .def("create_gather",
            [](TritonOpBuilder &self, Value src, Value indices, int axis)
                -> Value { return self.create<GatherOp>(src, indices, axis); })
+      .def("create_annotation",
+           [](TritonOpBuilder &self, Value &ptr, const std::string &attrKey,
+              Attribute &attrVal) {
+             auto annotationOp = self.create<AnnotationOp>(ptr);
+             annotationOp->setAttr(self.getBuilder().getStringAttr(attrKey),
+                                   attrVal);
+           })
       // Force GPU barrier
       .def("create_barrier",
            [](TritonOpBuilder &self) { self.create<mlir::gpu::BarrierOp>(); })
diff --git a/python/triton/language/semantic.py b/python/triton/language/semantic.py
index 0da22adef..dd08c8e14 100644
--- a/python/triton/language/semantic.py
+++ b/python/triton/language/semantic.py
@@ -1884,3 +1884,14 @@ class TritonSemantic(Generic[TensorTy]):
         handle = self.builder.create_make_tensor_descriptor(base_handle, [s.handle for s in shape],
                                                             [s.handle for s in strides], block_shape, is_signed_int)
         return tl.tensor_descriptor(handle, shape, strides, type)
+
+    def get_unit_attr(self):
+        return self.builder.get_unit_attr()
+    def get_bool_attr(self, value: bool):
+        return self.builder.get_bool_attr(value)
+    def get_int32_attr(self, value: int):
+        return self.builder.get_int32_attr(value)
+    def get_string_attr(self, value: str):
+        return self.builder.get_string_attr(value)
+    def create_annotation(self, ptr_handle, attr_key, attr_val):
+        return self.builder.create_annotation(ptr_handle, attr_key, attr_val)
\ No newline at end of file
