diff --git a/python/src/ir.cc b/python/src/ir.cc
old mode 100644
new mode 100755
index 387f73d1..82dac0d0
--- a/python/src/ir.cc
+++ b/python/src/ir.cc
@@ -38,8 +38,8 @@
 #include "llvm/Support/SourceMgr.h"
 
 #include "third_party/proton/dialect/include/Dialect/Proton/IR/Dialect.h"
+#include "third_party/spine_triton/include/triton-shared/Dialect/XSMT/IR/XSMTDialect.h"
 
-namespace {
 
 namespace py = pybind11;
 using namespace mlir;
@@ -47,6 +47,12 @@ using namespace triton;
 namespace tt = triton;
 namespace ttg = triton::gpu;
 namespace ttng = triton::nvidia_gpu;
+namespace ir {
+
+// Pointer to the TritonOpBuilder class, used to register IR ops for third-party
+// dialects.
+static py::class_<TritonOpBuilder> *builderClassPtr = nullptr;
+py::class_<TritonOpBuilder> *getBuilderClass() { return builderClassPtr; }
 
 llvm::raw_fd_ostream &mlir_dumps() {
   std::error_code EC;
@@ -190,13 +196,13 @@ py::list getTensorDescMetadata(ModuleOp &mod) {
   }
   return result;
 }
-
-} // anonymous namespace
+}
 
 /*****************************************************************************/
 /* Python bindings for ir                                                    */
 /*****************************************************************************/
 
+using namespace ir;
 void init_triton_ir(py::module &&m) {
   using ret = py::return_value_policy;
   using namespace pybind11::literals;
@@ -458,6 +464,7 @@ void init_triton_ir(py::module &&m) {
   py::class_<IntegerAttr, Attribute>(m, "integer_attr", py::module_local());
   py::class_<BoolAttr, Attribute>(m, "bool_attr", py::module_local());
   py::class_<UnitAttr, Attribute>(m, "unit_attr", py::module_local());
+  py::class_<StringAttr, Attribute>(m, "string_attr", py::module_local());
 
   // Ops
   py::class_<OpState>(m, "OpState", py::module_local())
@@ -711,9 +718,13 @@ void init_triton_ir(py::module &&m) {
 
   py::class_<OpBuilder::InsertPoint>(m, "InsertPoint", py::module_local());
 
-  py::class_<TritonOpBuilder>(m, "builder", py::module_local(),
-                              py::dynamic_attr())
-      .def(py::init<MLIRContext *>())
+  // The static builderClass object persists throughout the compilation,
+  // allowing third-party backends to register their ops separately.
+  static py::class_<TritonOpBuilder> builderClass(
+      m, "builder", py::module_local(), py::dynamic_attr());
+  builderClassPtr = &builderClass;
+  builderClass.def(py::init<MLIRContext *>())
+      .def("get_op_builder", &TritonOpBuilder::getBuilder, ret::reference)
       // getters
       .def("create_module",
            [](TritonOpBuilder &self) -> ModuleOp {
@@ -1717,6 +1728,11 @@ void init_triton_ir(py::module &&m) {
       .def("create_proton_record",
            [](TritonOpBuilder &self, bool isStart, int32_t regionId) -> void {
              self.create<mlir::triton::proton::RecordOp>(isStart, regionId);
+           })
+      .def("create_smt_bind",
+           [](TritonOpBuilder &self, Value &dest,
+              Value &src) -> Value {
+             return self.create<mlir::xsmt::BindOp>(dest.getType(), dest, src);
            });
 
   py::class_<PassManager>(m, "pass_manager", py::module_local())
diff --git a/python/src/ir.h b/python/src/ir.h
old mode 100644
new mode 100755
index b27c41a7..2e4d3b31
--- a/python/src/ir.h
+++ b/python/src/ir.h
@@ -2,6 +2,9 @@
 #include "mlir/IR/Builders.h"
 #include "triton/Tools/Sys/GetEnv.hpp"
 #include <memory>
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
 
 // A custom op builder that keeps track of the last location
 class TritonOpBuilder {
@@ -88,3 +91,7 @@ private:
   bool lineInfoEnabled =
       !mlir::triton::tools::getBoolEnv("TRITON_DISABLE_LINE_INFO");
 };
+
+namespace ir {
+extern py::class_<TritonOpBuilder> *getBuilderClass();
+} // namespace ir
\ No newline at end of file
diff --git a/python/triton/compiler/code_generator.py b/python/triton/compiler/code_generator.py
old mode 100644
new mode 100755
index 1dd7b4ed..68e4eb1c
--- a/python/triton/compiler/code_generator.py
+++ b/python/triton/compiler/code_generator.py
@@ -1036,8 +1036,12 @@ class CodeGenerator(ast.NodeVisitor):
         assert isinstance(node.ctx, ast.Store)
         lhs = self.visit(node.value)
         slices = self.visit(node.slice)
-        assert isinstance(lhs, language.tuple)
-        lhs.__setitem__(slices, value)
+        if isinstance(lhs, language.tensor):
+            lhs.__setitem__(slices, value, _semantic=self.semantic)
+        elif isinstance(lhs, language.tuple):
+            lhs.__setitem__(slices, value)
+        else:
+            raise RuntimeError('Only `tuple` and `tensor` are currently supported')
 
     def visit_Subscript(self, node):
         return self.visit_Subscript_Load(node)
@@ -1063,7 +1067,8 @@ class CodeGenerator(ast.NodeVisitor):
         disallow_acc_multi_buffer = False
         flatten = False
         warp_specialize = False
-        if IteratorClass is language.range:
+        bind_sub_block = None
+        if IteratorClass in [language.range, language.extra.smt.parallel]:
             iterator = IteratorClass(*iter_args, **iter_kwargs)
             # visit iterator arguments
             # note: only `range` iterator is supported now
@@ -1076,6 +1081,8 @@ class CodeGenerator(ast.NodeVisitor):
             disallow_acc_multi_buffer = iterator.disallow_acc_multi_buffer
             flatten = iterator.flatten
             warp_specialize = iterator.warp_specialize
+            if (IteratorClass is language.extra.smt.parallel):
+                bind_sub_block = iterator.bind_sub_block
         elif IteratorClass is range:
             # visit iterator arguments
             # note: only `range` iterator is supported now
@@ -1132,6 +1139,8 @@ class CodeGenerator(ast.NodeVisitor):
                 for_op.set_attr("tt.flatten", self.builder.get_unit_attr())
             if warp_specialize:
                 for_op.set_attr("tt.warp_specialize", self.builder.get_unit_attr())
+            if (bind_sub_block is not None) and bind_sub_block:
+                for_op.set_attr("bind_sub_block", self.builder.get_bool_attr(bind_sub_block))
 
             self.scf_stack.append(node)
             for_op_body = for_op.get_body(0)
diff --git a/python/triton/language/core.py b/python/triton/language/core.py
index 0f6030c1..cb3bd574 100644
--- a/python/triton/language/core.py
+++ b/python/triton/language/core.py
@@ -1118,6 +1118,12 @@ class tensor(base_value):
                 raise ValueError(f"unsupported tensor index: {sl}")
         return ret
 
+    @builtin
+    def __setitem__(self, slices, value, _semantic=None):
+        if slices is Ellipsis or slices == Ellipsis:
+            return _semantic.smt_bind(self, value)
+        raise ValueError(f"Tensor __setitem__ only supports [...] (Ellipsis), got {slices}")
+
     @property
     def T(self):
         """Transposes a 2D tensor."""
diff --git a/python/triton/language/semantic.py b/python/triton/language/semantic.py
index 0da22ade..9762df9c 100644
--- a/python/triton/language/semantic.py
+++ b/python/triton/language/semantic.py
@@ -1884,3 +1884,5 @@ class TritonSemantic(Generic[TensorTy]):
         handle = self.builder.create_make_tensor_descriptor(base_handle, [s.handle for s in shape],
                                                             [s.handle for s in strides], block_shape, is_signed_int)
         return tl.tensor_descriptor(handle, shape, strides, type)
+    def smt_bind(self, dest: tl.tensor, src: tl.tensor):
+        return self.tensor(self.builder.create_smt_bind(src.handle, dest.handle), dest.type)
diff --git a/setup.py b/setup.py
old mode 100644
new mode 100755
index 345986fa..2d9da394
--- a/setup.py
+++ b/setup.py
@@ -55,6 +55,8 @@ class Backend:
     tools_dir: Optional[str]
     install_dir: str
     is_external: bool
+    bin_dir: Optional[str]
+    lib_dir: Optional[str]
 
 
 class BackendInstaller:
@@ -88,13 +90,40 @@ class BackendInstaller:
         if not os.path.exists(tools_dir):
             tools_dir = None
 
+        bin_dir = os.path.join(backend_src_dir, "backend", "bin")
+        if not os.path.exists(bin_dir):
+            bin_dir = None
+        elif backend_name == "spine_triton":
+            spine_mlir_install_dir = os.environ.get("SPINE_MLIR_INSTALL_DIR", None)
+            assert spine_mlir_install_dir is not None
+            bin_install_list = ["llc", "spine-opt", "mlir-translate"]
+            for bin_file in bin_install_list:
+                shutil.copy(
+                    os.path.join(spine_mlir_install_dir, "bin", bin_file),
+                    os.path.join(bin_dir, bin_file)
+                )
+
+        lib_dir = os.path.join(backend_src_dir, "backend", "lib")
+        if not os.path.exists(lib_dir):
+            lib_dir = None
+        elif backend_name == "spine_triton":
+            spine_mlir_install_dir = os.environ.get("SPINE_MLIR_INSTALL_DIR", None)
+            assert spine_mlir_install_dir is not None
+            import glob
+            so_file_list = glob.glob(os.path.join(spine_mlir_install_dir, "lib", "libSpeIR*.so"))
+            for so_file in so_file_list:
+                shutil.copy(
+                    so_file,
+                    os.path.join(lib_dir, os.path.basename(so_file))
+                )
+
         for file in ["compiler.py", "driver.py"]:
             assert os.path.exists(os.path.join(backend_path, file)), f"${file} does not exist in ${backend_path}"
 
         install_dir = os.path.join(os.path.dirname(__file__), "python", "triton", "backends", backend_name)
 
         return Backend(name=backend_name, src_dir=backend_src_dir, backend_dir=backend_path, language_dir=language_dir,
-                       tools_dir=tools_dir, install_dir=install_dir, is_external=is_external)
+                       tools_dir=tools_dir, install_dir=install_dir, is_external=is_external, bin_dir=bin_dir, lib_dir=lib_dir)
 
     # Copy all in-tree backends under triton/third_party.
     @staticmethod
@@ -520,6 +549,14 @@ class CMakeBuild(build_ext):
         subprocess.check_call(["cmake", "--build", "."] + build_args, cwd=cmake_dir)
         subprocess.check_call(["cmake", "--build", ".", "--target", "mlir-doc"], cwd=cmake_dir)
 
+        for backend in backends:
+            if backend.name == "spine_triton":
+                spine_triton_opt_path = os.path.join(cmake_dir, "third_party", "spine_triton", "tools", "spine-triton-opt", "spine-triton-opt")
+                shutil.copy(
+                    spine_triton_opt_path,
+                    os.path.join(backend.bin_dir, "spine-triton-opt")
+                )
+
 
 def download_and_copy_dependencies():
     return
@@ -643,6 +680,12 @@ def get_packages():
             for x in os.listdir(backend.tools_dir):
                 yield f"triton.tools.extra.{x}"
 
+        if backend.bin_dir:
+            yield f"triton.backends.{backend.name}.bin"
+
+        if backend.lib_dir:
+            yield f"triton.backends.{backend.name}.lib"
+
     if check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
         yield "triton.profiler"
 
@@ -766,7 +809,7 @@ def get_git_version_suffix():
 
 
 # keep it separate for easy substitution
-TRITON_VERSION = "3.3.0" + get_git_version_suffix() + os.environ.get("TRITON_WHEEL_VERSION_SUFFIX", "")
+TRITON_VERSION = "3.3.0" + "+" + os.environ.get("SPINE_TRITON_VERSION_NUMBER", "")
 
 # Dynamically define supported Python versions and classifiers
 MIN_PYTHON = (3, 9)
@@ -790,7 +833,7 @@ setup(
     author="Philippe Tillet",
     author_email="phil@openai.com",
     description="A language and compiler for custom Deep Learning operations",
-    long_description="",
+    long_description="text/markdown",
     install_requires=[
         "setuptools>=40.8.0",
         "importlib-metadata; python_version < '3.10'",
